lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,378 @@
+
+void GrandCentralVisitor::visitDecl(WireOp op) { handleRef(op); };
+
+void GrandCentralVisitor::visitDecl(NodeOp op) { handleRef(op); };
+
+ArrayAttr GrandCentralVisitor::handleRefLike(mlir::Operation *op,
+                                             ArrayAttr annotations,
+                                             FIRRTLType type) {
+  if (!annotations)
+    return {};
+
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+
+  for (auto a : annotations) {
+    auto anno = a.cast<DictionaryAttr>();
+    auto clazz = anno.getAs<StringAttr>("class");
+
+    if (clazz == options->augmentedGroundType) {
+      auto defName = anno.getAs<StringAttr>("defName").getValue();
+      auto name = anno.getAs<StringAttr>("name").getValue();
+
+      // TODO: This is ignoring situations where the leaves of and interface are
+      // not ground types.  This enforces the requirement that this runs after
+      // LowerTypes.  However, this could eventually be relaxed.
+      if (!type.isGround()) {
+        // TODO: This diagnostic doesn't print as report_fatal_error causes a
+        // very quick exit. Fix this.
+        auto diag = op->emitOpError()
+                    << "cannot be added to interface '" << defName
+                    << "', component '" << name
+                    << "' because it is not a ground type. (Got type '" << type
+                    << "'.) This will be dropped from the interface. (Did you "
+                       "forget to run LowerTypes?)";
+        diag.attachNote()
+            << "The annotation indicating that this should be added was: '"
+            << a;
+        llvm::report_fatal_error(
+            "Grand Central needs to run after Lower Types. A "
+            "GroundTypeAnnotation was found on a non-ground type.");
+      }
+
+      auto component = &(*interfaceMap)[{defName, name}];
+      component->found |= true;
+
+      switch (component->tpe) {
+      case Component::Type::Bundle:
+        // Do nothing.
+        break;
+      case Component::Type::Vector:
+        component->width = type.getBitWidthOrSentinel();
+        component->depth++;
+        break;
+      case Component::Type::Ground:
+        component->width = type.getBitWidthOrSentinel();
+        break;
+      case Component::Type::Unknown:
+        llvm_unreachable("Shouldn't be here");
+      }
+
+      continue;
+    }
+
+    unprocessedAnnos.push_back(a);
+  }
+
+  if (unprocessedAnnos.empty())
+    return {};
+  else
+    return ArrayAttr::get(context, unprocessedAnnos);
+};
+
+/// Combined logic to handle both Wires, Registers, and Nodes because these all
+/// use the same approach.
+void GrandCentralVisitor::handleRef(mlir::Operation *op) {
+
+  auto remainingAnnotations = handleRefLike(
+      op, getAnnotations(op), op->getResult(0).getType().cast<FIRRTLType>());
+
+  if (!remainingAnnotations || remainingAnnotations.empty())
+    op->removeAttr("annotations");
+  else
+    op->setAttr("annotations", remainingAnnotations);
```

Comment:
I believe that $annotations is always a required attribute, so setting it to an empty array is the right way to go.  You can remove the "removeAttr" case to simplify this.

Reply:
Great observation. The functions that return remaining `ArrayAttr` will now return an empty array instead of a null attribute. That cleans this up and the annotations can be blindly overwritten.